### PR TITLE
extension tests and added strict mode all over

### PR DESCRIPTION
**Pull request by crankycoder**
- I've forced all code and tests to "use strict"
- added two tests for extension validation slight modifications to code to fire errors for bad keys or bad values in extensions

### CODE REVIEW

File: lib/cef.js [hunk 1]
```diff
@@ -11,19 +11,20 @@
  *
  * Please see the README.md file for usage.
  */
+"use strict";
 
-const os = require('os');
-const syslog = require('./syslog');
-const Formatter = require('./formatter');
```
Comment (jedp) [inline lib/cef.js:17]: Why the change from const to var, here and elsewhere?  Is it because of `strict` mode?  Maybe this is heretical, but I prefer having `const` to `strict`.

File: lib/formatter.js [hunk 2]
```diff
@@ -1,8 +1,38 @@
...
+ * This module provides a logger that formats event data according to
+ * the CEF standard and, by default, emits them to the syslog over
+ * udp.  The logger backend is pluggable and can be changed if this
+ * does not suit your needs.
+ *
+ * Please see the README.md file for usage.
+ */
+"use strict";
+
+var _ = require('underscore');
```
Comment (jedp) [inline lib/formatter.js]: I'd rather not introduce another dependency.  Is underscore necessary?

File: lib/formatter.js [hunk 3]
```diff
@@ -147,19 +177,27 @@ Formatter.prototype = {
     var extensionArray = [];
     var value = "";
     var validator = null;
-    Object.keys(extensions).forEach(function(key) {
+    var err = null;
+    _.each(_.keys(extensions), function(key) {
```
Comment (jedp) [inline lib/formatter.js]: For example, here I don't see what motivates the change from `Object.keys(...).forEach(...)` to the underscore methods.

File: lib/formatter.js [hunk 4]
```diff
@@ -147,19 +177,27 @@ Formatter.prototype = {
...
       validator = validatorForKey(key);
       if (validator) {
         value = extensions[key];
         if (validator(value)) {
           extensionArray.push(util.format("%s=%s", key, this.sanitizeExtensionValue(value)));
         } else {
-          console.error(util.format("Not a valid value for %s: %s", key, value));
+          err = new Error(util.format("Not a valid value for %s: %s", key, value));
+          return err;
```
Comment (jedp) [inline lib/formatter.js:186]: So the return here would just return from the inner function arg to `each`, which seems like it would not do anything, would it?  The way I had it before was not perfect, but it would stash the error to be returned after the iteration was complete.  (It could maybe have been improved by keeping a list of errors and returning them all?)

However, I'm definitely open to discussing how this method behaves on error.  My thinking before was that the formatter should not throw an exception if something was invalid, but that it should simply format what it could.  (Though looking at it now, I don't think `console.error` is the way to go about it.)  I can envision cases where an engineer tried to send, say, a string, but accidentally sent a json object or something, in which case all the events would be lost.  Put another way, if you tell it to log an event, I think it should log everything it can.  What do you think?

File: lib/formatter.js [hunk 5]
```diff
@@ -147,19 +177,27 @@ Formatter.prototype = {
     var extensionArray = [];
     var value = "";
     var validator = null;
-    Object.keys(extensions).forEach(function(key) {
+    var err = null;
```
Comment (jedp) [inline lib/formatter.js:177]: Good catch - looks like i was modifying a global `err` below.

File: tests/test_cef_formatter.js [hunk 6]
```diff
@@ -191,8 +194,64 @@ var suite = vows.describe("Formatter")
...
+                  extensions: {
+                    rt: "20130320",
+                  }
+                };
+                return formatter.format(params);
+            },
+
+            "throws an error": function(_, err, result) {
+                assert(err);
+                assert(!result);
```
Comment (jedp) [inline tests/test_cef_formatter.js:223]: See comments above; I'm not sure the formatter should completely fail on bad input.  I think it should be lenient so messages don't get thrown away completely just because one field is wrong.

**Comment by jedp** [comment]
Hi, @crankycoder, and thanks for the PR.  I'd like to work through the questions above before accepting this.  Would you reply inline?  Thanks!

File: lib/cef.js (hunk 1 above)
Comment (crankycoder) [inline lib/cef.js:17]: Yes, this is because of the use of `strict` mode.  This is mostly a habit of mine, I just push all languages into strict mode out of habit. Consider nodejs only runs on V8 and not just any javascript engine, I think the only real payoff here is well, more strictness.  I have a preference for throwing errors on bad syntax instead of javascript silently eating the error.

On the other hand, I only really care about adding heka transport support to node-cef, so if this is a deal breaker, I'll defer to your judgement.

File: lib/formatter.js (hunk 2 above)
Comment (crankycoder) [inline lib/formatter.js]: i'll remove underscore

File: lib/formatter.js (hunk 4 above)
Comment (crankycoder) [inline lib/formatter.js:186]: Thinking about this some more, I think I'd really rather have the errors propagate up to the caller.   heka supports raven and sentry for stacktrace support and I think it makes sense for CEF messages to fire off for discrete security related events.

In the case that the node-cef library itself is called incorrectly, I think we should hook into the raven client to send a stacktrace to Sentry so that we can see the error in the application.  If we really do have a case where a developer accidentally passes in the wrong type to the the CEF logger, I'd like to treat that as an actual error and have raven hand off a full stackframe to sentry for later analysis.

File: tests/test_cef_formatter.js (hunk 6 above)
Comment (crankycoder) [inline tests/test_cef_formatter.js:223]: See https://github.com/jedp/node-cef/pull/6#r3639989 , I'm leaning towards being intolerant of errors in the code and to just fail fast. We can pass off errors to a raven client and log them to sentry.  I think we have a better chance of getting debug information out of our apps with a full stacktrace.

**Comment by crankycoder** [comment]
Hi @jedp!  Thanks for the feedback, I've checked V8 and there is no actual performance benefit of having strict mode, so I've reverted the strict mode changes.

The underscore dependency has been removed now.

I'm still leaning towards having fail-fast behavior for invalidly formatted messages.  After heka+node-cef integration lands, I'm moving to getting raven/sentry+node working so that we have nice stack traces that will help debug problems when errors start to occur in a production setting.  I think separating "security events" from actual errors in the code makes sense.

**Comment by crankycoder** [comment]
I'm reworking this patch to meet you half way.  Closing this as I'll just file a new PR later.
